### PR TITLE
WIP: Standalone deb builds

### PR DESCRIPTION
**Pull request by ottok**
Build only ColumnStore and produce only the mariadb-plugin-columnstore binary package to be able to publish it independently of the server.

***DO NOT MERGE THIS***. This is just an experiment created as a PR to get the Drone CI run on it.

### CODE REVIEW

**Comment by ottok** [comment]
To finish this I would need to know what are the actual build dependencies of ColumnStore (https://jira.mariadb.org/browse/MCOL-4471) and what are the required end products? Would just one mariadb-plugin-columnstore.deb and .rpm suffice? Or do you also need to produce other parts of the server?

**Comment by ottok** [comment]
Currently the Drone CI spits out these extra packages in addition to the ColumnStore binary package:
```
-rw-r--r--    1 root     root          2964 Dec 27 21:03 libmariadb-dev-compat_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root          7404 Dec 27 21:03 libmariadb-dev-dbgsym_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root        159352 Dec 27 21:03 libmariadb-dev_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root          2796 Dec 27 21:03 libmariadb3-compat_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root        490990 Dec 27 21:03 libmariadb3-dbgsym_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root        148348 Dec 27 21:03 libmariadb3_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root          2762 Dec 27 21:03 libmariadbclient18_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root          2770 Dec 27 21:03 libmysqlclient18_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root       7604090 Dec 27 21:03 mariadb-client-10.6-dbgsym_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root       1506680 Dec 27 21:03 mariadb-client-10.6_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root       2499692 Dec 27 21:03 mariadb-client-core-10.6-dbgsym_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root        764264 Dec 27 21:03 mariadb-client-core-10.6_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root          2608 Dec 27 21:03 mariadb-client_10.6.0+maria~stretch_all.deb
-rw-r--r--    1 root     root          3726 Dec 27 21:03 mariadb-common_10.6.0+maria~stretch_all.deb
-rw-r--r--    1 root     root     185777294 Dec 27 21:03 mariadb-plugin-columnstore-dbgsym_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root       6665360 Dec 27 21:03 mariadb-plugin-columnstore_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root          8756 Dec 27 21:03 mariadb-plugin-cracklib-password-check-dbgsym_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root          4740 Dec 27 21:03 mariadb-plugin-cracklib-password-check_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root       2752700 Dec 27 21:03 mariadb-plugin-s3-dbgsym_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root        803312 Dec 27 21:03 mariadb-plugin-s3_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root      31692266 Dec 27 21:03 mariadb-server-10.6-dbgsym_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root       4170620 Dec 27 21:03 mariadb-server-10.6_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root      74647922 Dec 27 21:03 mariadb-server-core-10.6-dbgsym_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root       6200210 Dec 27 21:03 mariadb-server-core-10.6_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root          2732 Dec 27 21:03 mariadb-server_10.6.0+maria~stretch_all.deb
-rw-r--r--    1 root     root      17244516 Dec 27 21:03 mariadb-test-data_10.6.0+maria~stretch_all.deb
-rw-r--r--    1 root     root       4282484 Dec 27 21:03 mariadb-test-dbgsym_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root       1172146 Dec 27 21:03 mariadb-test_10.6.0+maria~stretch_amd64.deb
-rw-r--r--    1 root     root          2778 Dec 27 21:03 mysql-common_10.6.0+maria~stretch_all.deb
```

Is it really necessary to build all these? Would it not suffice to just build the ColumnStore package and then pull in the necessary MariaDB Server bits from the corresponding mariadb.org repository are prebuilt binary packages?

**Comment by mariadb-RomanNavrotskiy** [comment]
> Currently the Drone CI spits out these extra packages in addition to the ColumnStore binary package:
> 
> ```
> -rw-r--r--    1 root     root          2964 Dec 27 21:03 libmariadb-dev-compat_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root          7404 Dec 27 21:03 libmariadb-dev-dbgsym_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root        159352 Dec 27 21:03 libmariadb-dev_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root          2796 Dec 27 21:03 libmariadb3-compat_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root        490990 Dec 27 21:03 libmariadb3-dbgsym_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root        148348 Dec 27 21:03 libmariadb3_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root          2762 Dec 27 21:03 libmariadbclient18_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root          2770 Dec 27 21:03 libmysqlclient18_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root       7604090 Dec 27 21:03 mariadb-client-10.6-dbgsym_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root       1506680 Dec 27 21:03 mariadb-client-10.6_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root       2499692 Dec 27 21:03 mariadb-client-core-10.6-dbgsym_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root        764264 Dec 27 21:03 mariadb-client-core-10.6_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root          2608 Dec 27 21:03 mariadb-client_10.6.0+maria~stretch_all.deb
> -rw-r--r--    1 root     root          3726 Dec 27 21:03 mariadb-common_10.6.0+maria~stretch_all.deb
> -rw-r--r--    1 root     root     185777294 Dec 27 21:03 mariadb-plugin-columnstore-dbgsym_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root       6665360 Dec 27 21:03 mariadb-plugin-columnstore_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root          8756 Dec 27 21:03 mariadb-plugin-cracklib-password-check-dbgsym_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root          4740 Dec 27 21:03 mariadb-plugin-cracklib-password-check_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root       2752700 Dec 27 21:03 mariadb-plugin-s3-dbgsym_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root        803312 Dec 27 21:03 mariadb-plugin-s3_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root      31692266 Dec 27 21:03 mariadb-server-10.6-dbgsym_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root       4170620 Dec 27 21:03 mariadb-server-10.6_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root      74647922 Dec 27 21:03 mariadb-server-core-10.6-dbgsym_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root       6200210 Dec 27 21:03 mariadb-server-core-10.6_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root          2732 Dec 27 21:03 mariadb-server_10.6.0+maria~stretch_all.deb
> -rw-r--r--    1 root     root      17244516 Dec 27 21:03 mariadb-test-data_10.6.0+maria~stretch_all.deb
> -rw-r--r--    1 root     root       4282484 Dec 27 21:03 mariadb-test-dbgsym_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root       1172146 Dec 27 21:03 mariadb-test_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root          2778 Dec 27 21:03 mysql-common_10.6.0+maria~stretch_all.deb
> ```
> 
> Is it really necessary to build all these? Would it not suffice to just build the ColumnStore package and then pull in the necessary MariaDB Server bits from the corresponding mariadb.org repository are prebuilt binary packages?

because of the package versions conflicts.

packages from develop branch built against HEAD of 10.6 mariadb server repo. mariadb server prebuilt binary packages would have conflicts most of the time.

**Comment by ottok** [comment]
> because of the package versions conflicts

Can you please elaborate on this? If the issue is simply package metadata in .deb, we can fix it easily.

I am interested here in what are the actual functional dependencies/relations?

**Comment by mariadb-RomanNavrotskiy** [comment]
> > because of the package versions conflicts
> 
> Can you please elaborate on this? If the issue is simply package metadata in .deb, we can fix it easily.
> 
> I am interested here in what are the actual functional dependencies/relations?

you can try it by yourself. just add official mariadb repositories and try to install columnstore packages from some of recent develop branch builds. and yes, this is simply package dependencies problem (metadata in .deb). so if you can fix it easly – you are welcome!

**Comment by ottok** [comment]
> you can try it by yourself. just add official mariadb repositories and try to install columnstore packages from some of recent develop branch builds. and yes, this is simply package dependencies problem (metadata in .deb). so if you can fix it easly – you are welcome!

That was not the point. I am asking you what the dependency is? I am quite sure putting any `usr/lib/mysql/plugin/ha_columnstore.so` with any version of the MariaDB Server would not work.

The metadata in the deb exists only to reflect what the dependency is. Now you are telling me that there is no dependency and the version strings in the .deb packaging are artificial. Surely there is a real dependency and you as ColumnStore developers should know it.

**Comment by mariadb-RomanNavrotskiy** [comment]
> > you can try it by yourself. just add official mariadb repositories and try to install columnstore packages from some of recent develop branch builds. and yes, this is simply package dependencies problem (metadata in .deb). so if you can fix it easly – you are welcome!
> 
> That was not the point. I am asking you what the dependency is? I am quite sure putting any `usr/lib/mysql/plugin/ha_columnstore.so` with any version of the MariaDB Server would not work.
> 
> The metadata in the deb exists only to reflect what the dependency is. Now you are telling me that there is no dependency and the version strings in the .deb packaging are artificial. Surely there is a real dependency and you as ColumnStore developers should know it.

I talk in context of you initial comment about 'extra packages'. Users usually get packages, not the ha_columnstore.so file.

We have no independent build system and build process (from mariadb server project). We just build server with columnstore plugin. Artefacts of the build are packages. So, feel free to rework all the build process (from binaries to packages)!

As for real dependency from server I know nothing about that. As far I do only CI part. I think plugin interface should be compatible across major (or minor?) versions. Let's wait for answers from 'real' developers like @drrtuy about that.

**Comment by drrtuy** [comment]
> Currently the Drone CI spits out these extra packages in addition to the ColumnStore binary package:
> 
> ```
> -rw-r--r--    1 root     root          2964 Dec 27 21:03 libmariadb-dev-compat_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root          7404 Dec 27 21:03 libmariadb-dev-dbgsym_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root        159352 Dec 27 21:03 libmariadb-dev_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root          2796 Dec 27 21:03 libmariadb3-compat_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root        490990 Dec 27 21:03 libmariadb3-dbgsym_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root        148348 Dec 27 21:03 libmariadb3_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root          2762 Dec 27 21:03 libmariadbclient18_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root          2770 Dec 27 21:03 libmysqlclient18_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root       7604090 Dec 27 21:03 mariadb-client-10.6-dbgsym_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root       1506680 Dec 27 21:03 mariadb-client-10.6_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root       2499692 Dec 27 21:03 mariadb-client-core-10.6-dbgsym_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root        764264 Dec 27 21:03 mariadb-client-core-10.6_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root          2608 Dec 27 21:03 mariadb-client_10.6.0+maria~stretch_all.deb
> -rw-r--r--    1 root     root          3726 Dec 27 21:03 mariadb-common_10.6.0+maria~stretch_all.deb
> -rw-r--r--    1 root     root     185777294 Dec 27 21:03 mariadb-plugin-columnstore-dbgsym_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root       6665360 Dec 27 21:03 mariadb-plugin-columnstore_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root          8756 Dec 27 21:03 mariadb-plugin-cracklib-password-check-dbgsym_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root          4740 Dec 27 21:03 mariadb-plugin-cracklib-password-check_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root       2752700 Dec 27 21:03 mariadb-plugin-s3-dbgsym_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root        803312 Dec 27 21:03 mariadb-plugin-s3_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root      31692266 Dec 27 21:03 mariadb-server-10.6-dbgsym_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root       4170620 Dec 27 21:03 mariadb-server-10.6_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root      74647922 Dec 27 21:03 mariadb-server-core-10.6-dbgsym_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root       6200210 Dec 27 21:03 mariadb-server-core-10.6_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root          2732 Dec 27 21:03 mariadb-server_10.6.0+maria~stretch_all.deb
> -rw-r--r--    1 root     root      17244516 Dec 27 21:03 mariadb-test-data_10.6.0+maria~stretch_all.deb
> -rw-r--r--    1 root     root       4282484 Dec 27 21:03 mariadb-test-dbgsym_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root       1172146 Dec 27 21:03 mariadb-test_10.6.0+maria~stretch_amd64.deb
> -rw-r--r--    1 root     root          2778 Dec 27 21:03 mysql-common_10.6.0+maria~stretch_all.deb
> ```
> 
> Is it really necessary to build all these? Would it not suffice to just build the ColumnStore package and then pull in the necessary MariaDB Server bits from the corresponding mariadb.org repository are prebuilt binary packages?

That was my original idea but the actual implementors of the rubuilding process prefered to go the other way.

**Comment by ottok** [comment]
Maybe that route was taken due to the lack of documentation of *actual*
dependencies?

If you publish a list somewhere (e.g.
https://jira.mariadb.org/browse/MCOL-4471) then I could engineer the
build + Debian packaging to produce only the ColumnStore itself and omit
e.g. s3 plugin and other cruft that is surely not needed.

**Comment by drrtuy** [comment]
> Maybe that route was taken due to the lack of documentation of *actual* dependencies? If you publish a list somewhere (e.g. https://jira.mariadb.org/browse/MCOL-4471) then I could engineer the build + Debian packaging to produce only the ColumnStore itself and omit e.g. s3 plugin and other cruft that is surely not needed.
Maybe.
The packages needed to compile MCS are available [here](https://github.com/mariadb-corporation/mariadb-columnstore-engine/blob/develop/install-deps.sh).

**Comment by drrtuy** [comment]
I believe we had merged the idea with another patch of yours so closing this one. Plz msg me if I am wrong @ottok .

**Comment by ottok** [comment]
This is a separate idea/PR, please don't close it.

**Comment by ottok** [comment]
@drrtuy How do I trigger the CI for this PR? None seem to have run now after I force pushed an improved version of this.

**Comment by ottok** [comment]
The general structure is correct, but builds fail on some pthread issue maybe due to missing dependency?

https://ci.columnstore.mariadb.net/mariadb-corporation/mariadb-columnstore-engine/2043/7/4
> /usr/bin/cc -g -O2 -fdebug-prefix-map=/mdb/verylongdirnameforverystrangecpackbehavior=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wdate-time -D_FORTIFY_SOURCE=2 -pie -fPIC -fstack-protector --param=ssp-buffer-size=4 -Wunused -Wlogical-op -Wno-uninitialized -Wall -Wextra -Wformat-security -Wno-init-self -Wwrite-strings -Wshift-count-overflow -Wdeclaration-after-statement -Wno-undef -Wno-unknown-pragmas -DCHECK_FUNCTION_EXISTS=pthread_getspecific  -Wl,-Bsymbolic-functions -Wl,-z,relro -Wl,-z,now  -rdynamic CMakeFiles/cmTC_8455f.dir/CheckFunctionExists.c.o  -o cmTC_8455f 
12097	CMakeFiles/cmTC_8455f.dir/CheckFunctionExists.c.o: In function `main':
12098	/usr/share/cmake-3.10/Modules/CheckFunctionExists.c:17: undefined reference to `pthread_getspecific'
12099	collect2: error: ld returned 1 exit status

File: debian/control [hunk 1]
```diff
@@ -1,3 +1,61 @@
...
+               libncurses5-dev (>= 5.0-6~),
+               libncurses5-dev:native (>= 5.0-6~),
+               libnuma-dev [linux-any],
+               libpam0g-dev,
+               libpcre2-dev,
+               libsnappy-dev,
+               libssl-dev,
+               libssl-dev:native,
+               libsystemd-dev [linux-any],
+               libaio-dev [linux-any],
```
Comment (drrtuy) [inline debian/control:44]: We don't require libaio-dev yet.

Comment (ottok) [reply debian/control:44]: Ok, removed libaio-dev

**Comment by ottok** [comment]
@drrtuy Build currently failing in CI on these – any idea why?

Some customization to build dependencies or CMake command needed?
```
Building CXX object CMakeFiles/cmTC_a2ff9.dir/src.cxx.o
/usr/bin/c++  -DPACKAGE=test -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE=1 -D_LARGEFILE_SOURCE -D_LARGE_FILES -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS  -g -O2 -fdebug-prefix-map=/mdb/verylongdirnameforverystrangecpackbehavior=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wdate-time -D_FORTIFY_SOURCE=2 -pie -fPIC -fstack-protector --param=ssp-buffer-size=4 -Dhave_CXX__Wunused_private_field  -Wunused-private-field   -o CMakeFiles/cmTC_a2ff9.dir/src.cxx.o -c /mdb/verylongdirnameforverystrangecpackbehavior/builddir/CMakeFiles/CMakeTmp/src.cxx
c++: error: unrecognized command line option '-Wunused-private-field'
make[3]: *** [CMakeFiles/cmTC_a2ff9.dir/build.make:66: CMakeFiles/cmTC_a2ff9.dir/src.cxx.o] Error 1
make[3]: Leaving directory '/mdb/verylongdirnameforverystrangecpackbehavior/builddir/CMakeFiles/CMakeTmp'
make[2]: *** [Makefile:121: cmTC_a2ff9/fast] Error 2
...
Building C object CMakeFiles/cmTC_f36a6.dir/SIZEOF_SOCKLEN_T.c.o
/usr/bin/cc -DPACKAGE=test -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE=1 -D_LARGEFILE_SOURCE -D_LARGE_FILES -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS  -g -O2 -fdebug-prefix-map=/mdb/verylongdirnameforverystrangecpackbehavior=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wdate-time -D_FORTIFY_SOURCE=2 -pie -fPIC -fstack-protector --param=ssp-buffer-size=4 -Wunused -Wlogical-op -Wno-uninitialized -Wall -Wextra -Wformat-security -Wno-init-self -Wwrite-strings -Wshift-count-overflow -Wdeclaration-after-statement -Wno-undef -Wno-unknown-pragmas    -o CMakeFiles/cmTC_f36a6.dir/SIZEOF_SOCKLEN_T.c.o   -c /mdb/verylongdirnameforverystrangecpackbehavior/builddir/CMakeFiles/CheckTypeSize/SIZEOF_SOCKLEN_T.c
/mdb/verylongdirnameforverystrangecpackbehavior/builddir/CMakeFiles/CheckTypeSize/SIZEOF_SOCKLEN_T.c:24:22: error: 'socklen_t' undeclared here (not in a function)
   24 | #define SIZE (sizeof(socklen_t))
      |                      ^~~~~~~~~
/mdb/verylongdirnameforverystrangecpackbehavior/builddir/CMakeFiles/CheckTypeSize/SIZEOF_SOCKLEN_T.c:26:12: note: in expansion of macro 'SIZE'
   26 |   ('0' + ((SIZE / 10000)%10)),
      |            ^~~~
make[3]: *** [CMakeFiles/cmTC_f36a6.dir/build.make:66: CMakeFiles/cmTC_f36a6.dir/SIZEOF_SOCKLEN_T.c.o] Error 1
make[3]: Leaving directory '/mdb/verylongdirnameforverystrangecpackbehavior/builddir/CMakeFiles/CMakeTmp'
make[2]: *** [Makefile:121: cmTC_f36a6/fast] Error 2
...
/usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_11fb8.dir/link.txt --verbose=1
/usr/bin/cc -g -O2 -fdebug-prefix-map=/mdb/verylongdirnameforverystrangecpackbehavior=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wdate-time -D_FORTIFY_SOURCE=2 -pie -fPIC -fstack-protector --param=ssp-buffer-size=4 -Wunused -Wlogical-op -Wno-uninitialized -Wall -Wextra -Wformat-security -Wno-init-self -Wwrite-strings -Wshift-count-overflow -Wdeclaration-after-statement -Wno-undef -Wno-unknown-pragmas -DCHECK_FUNCTION_EXISTS=floor  -Wl,-Bsymbolic-functions -Wl,-z,relro -Wl,-z,now  -rdynamic CMakeFiles/cmTC_11fb8.dir/CheckFunctionExists.c.o  -o cmTC_11fb8 
/usr/bin/ld: CMakeFiles/cmTC_11fb8.dir/CheckFunctionExists.c.o: in function `main':
/usr/share/cmake-3.16/Modules/CheckFunctionExists.c:17: undefined reference to `floor'
collect2: error: ld returned 1 exit status
...
/usr/bin/cc -g -O2 -fdebug-prefix-map=/mdb/verylongdirnameforverystrangecpackbehavior=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wdate-time -D_FORTIFY_SOURCE=2 -pie -fPIC -fstack-protector --param=ssp-buffer-size=4 -Wunused -Wlogical-op -Wno-uninitialized -Wall -Wextra -Wformat-security -Wno-init-self -Wwrite-strings -Wshift-count-overflow -Wdeclaration-after-statement -Wno-undef -Wno-unknown-pragmas -DCHECK_FUNCTION_EXISTS=pthread_getspecific  -Wl,-Bsymbolic-functions -Wl,-z,relro -Wl,-z,now  -rdynamic CMakeFiles/cmTC_96279.dir/CheckFunctionExists.c.o  -o cmTC_96279 
/usr/bin/ld: CMakeFiles/cmTC_96279.dir/CheckFunctionExists.c.o: in function `main':
/usr/share/cmake-3.16/Modules/CheckFunctionExists.c:17: undefined reference to `pthread_getspecific'
collect2: error: ld returned 1 exit status
```

**Comment by ottok** [comment]
Ok, now build is passing on most Debian/Ubuntu platforms.  If run-time dependency `mariadb-server-10.5 (= ${server:Version}),` the smoke tests will fail to start with:
```
The following packages have unmet dependencies:
 mariadb-plugin-columnstore : Depends: mariadb-server-10.5 (= 1:10.5.10+columnstore~bionic)
 ```
 
 On Ubuntu 16.04 "Xenial" build fails due to missing libmariadb-dev, but that can be ignored as Xenial goes EOL in April 2021.

Next step would be to either wait for https://jira.mariadb.org/browse/MCOL-4471 to be done, or test it by brute force doing 43 builds, each with one dependency dropped. We could also review all warnings such as `dpkg-shlibdeps: warning: package could avoid a useless dependency ...were not linked against libdl.so.2 (they use none of the library's symbols)` as a source of information on what dependencies to drop (although this example with libdl.so.2. is not good as it is from libc6, which is surely needed).

**Comment by ottok** [comment]
Also I will also test `cmake --build . --target columnstore`. I am taking a break from this now but will return later to do a new round of work on this.

**Comment by ottok** [comment]
@drrtuy @mariadb-RomanNavrotskiy This is almost done, but I have some debugging commits as I am trying to understand why does the ColumnStore service fail to start up? The smoke tests are not passing because of this. Example: https://ci.columnstore.mariadb.net/mariadb-corporation/mariadb-columnstore-engine/2249/7/7
```
docker exec -t smoke${DRONE_BUILD_NUMBER} systemctl status --no-pager mariadb-columnstore
951	● mariadb-columnstore.service - mariadb-columnstore
952	     Loaded: loaded (; enabled; vendor preset: enabled)
953	     Active: active (exited) since Tue 2021-04-27 05:04:45 UTC; 89ms ago
954	    Process: 5957 ExecStart=/usr/bin/mariadb-columnstore-start.sh (code=exited, status=0/SUCCESS)
955	   Main PID: 5957 (code=exited, status=0/SUCCESS)
957	Apr 27 05:04:45 69222369743c systemd[1]: Starting mariadb-columnstore...
958	Apr 27 05:04:45 69222369743c su[6095]: (to mysql) root on none
959	Apr 27 05:04:45 69222369743c su[6095]: pam_env(su:session): Unable to open …tory
960	Apr 27 05:04:45 69222369743c su[6095]: pam_unix(su:session): session opened…d=0)
961	Apr 27 05:04:45 69222369743c writeengine[6097]: 45.814318 |0|0|0| I 19 CAL00…ed.
962	Apr 27 05:04:45 69222369743c su[6095]: pam_unix(su:session): session closed…ysql
963	Apr 27 05:04:45 69222369743c systemd[1]: Finished mariadb-columnstore.
```

**Comment by ottok** [comment]
drrtuy on Zulip:
> JFYI systemd unit mariadb-columnstore is just a oneshoot service that runs /usr/bin/mariadb-columnstore-start.sh so that the status you mentioned looks normal to me. So the next step would be log inspection that I can do when you rebase and update the branch.

**Comment by ottok** [comment]
@drrtuy Seems ColumnStore is indeed installed (as it should) but using it kills the thread or server:
```
MariaDB [test]> show plugins soname where library like 'ha_columnstore.so';
+---------------------+--------+--------------------+-------------------+---------+
| Name                | Status | Type               | Library           | License |
+---------------------+--------+--------------------+-------------------+---------+
| Columnstore         | ACTIVE | STORAGE ENGINE     | ha_columnstore.so | GPL     |
| COLUMNSTORE_COLUMNS | ACTIVE | INFORMATION SCHEMA | ha_columnstore.so | GPL     |
| COLUMNSTORE_TABLES  | ACTIVE | INFORMATION SCHEMA | ha_columnstore.so | GPL     |
| COLUMNSTORE_FILES   | ACTIVE | INFORMATION SCHEMA | ha_columnstore.so | GPL     |
| COLUMNSTORE_EXTENTS | ACTIVE | INFORMATION SCHEMA | ha_columnstore.so | GPL     |
+---------------------+--------+--------------------+-------------------+---------+
5 rows in set (0.045 sec)

MariaDB [test]> show create table t1;
+-------+---------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                |
+-------+---------------------------------------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (
  `a` int(11) DEFAULT NULL
) ENGINE=Columnstore DEFAULT CHARSET=utf8mb4 |
+-------+---------------------------------------------------------------------------------------------+
1 row in set (0.000 sec)

MariaDB [test]> insert into test.t1 values (2); select * from test.t1;
ERROR 2006 (HY000): MySQL server has gone away
No connection. Trying to reconnect...
Connection id:    23
Current database: test

ERROR 2013 (HY000): Lost connection to MySQL server during query
ERROR 2006 (HY000): MySQL server has gone away
No connection. Trying to reconnect...
ERROR 2002 (HY000): Can't connect to local MySQL server through socket '/run/mysqld/mysqld.sock' (111)
ERROR: Can't connect to the server

unknown [test]> show errors;
No connection. Trying to reconnect...
Connection id:    23
Current database: test

Empty set (0.000 sec)

MariaDB [test]> show warnings;
Empty set (0.000 sec)
```
How do you recommend I debug this failure? Where will ColumnStore log something when it dies?

**Comment by ottok** [comment]
@drrtuy This might be the root cause:
```
# mariadb -e "INSTALL PLUGIN Columnstore SONAME 'ha_columnstore.so';"
ERROR 1126 (HY000) at line 1: Can't open shared library 'ha_columnstore.so' (errno: 8, API version for STORAGE ENGINE plugin Columnstore not supported by this version of the server)
```

Where is the API version defined/compared?

On the server side seems to be `sql/sql_plugin.cc: " not supported by this version of the server", NullS);`

Is it in `ha_mcs.cpp`?
```
struct st_mysql_storage_engine ha_mcs_cache_storage_engine=
{ MYSQL_HANDLERTON_INTERFACE_VERSION };

struct st_mysql_storage_engine columnstore_storage_engine =
{ MYSQL_HANDLERTON_INTERFACE_VERSION };

static struct st_mysql_information_schema is_columnstore_plugin_version =
{ MYSQL_INFORMATION_SCHEMA_INTERFACE_VERSION };
```

There is no material dependency to the server, you simply define that you always depend on the same server as built with?

**Comment by ottok** [comment]
Seems the ColumnStore build itself is very slow, and trimming away the other parts don't help much in build speed. However, using ccache does speed up the build 90%. Tried integrating [drone-cache](http://plugins.drone.io/meltwater/drone-cache/) but can't complete it without your help, as it would need a new S3 bucket to store cache in:
```
level=error name=drone-cache ts=2021-05-23T20:43:54.270125703Z caller=main.go:466 err="[WARNING] build cache, process rebuild failed, upload upload file put the object NoSuchBucket: The specified bucket does not exist\n\tstatus code: 404, request id: HK5NZ5DCKE9VHRJT, host id: QfOewvsVfzFb8hsil1/tdyQk4T1bJPXLgT0czTa39Uwl3/9NwsxOXJRw91TFAAHOjxM00YJZmL0=\n"
```

**Comment by ottok** [comment]
After hacking the server version scheme ColumnStore does install, but write operations die:
```
+ docker exec -t smoke${DRONE_BUILD_NUMBER} mariadb -e "insert into test.t1 values (2); select * from test.t1"
ERROR 2013 (HY000) at line 1: Lost connection to MySQL server during query
```

I was not able to debug why.

**Comment by drrtuy** [comment]
> After hacking the server version scheme ColumnStore does install, but write operations die:
> 
> ```
> + docker exec -t smoke${DRONE_BUILD_NUMBER} mariadb -e "insert into test.t1 values (2); select * from test.t1"
> ERROR 2013 (HY000) at line 1: Lost connection to MySQL server during query
> ```
> 
> I was not able to debug why.

I will run the build and look at the server trace when MDB thread crashes. I will message you back with the results.

**Comment by drrtuy** [comment]
> Seems the ColumnStore build itself is very slow, and trimming away the other parts don't help much in build speed. However, using ccache does speed up the build 90%. Tried integrating [drone-cache](http://plugins.drone.io/meltwater/drone-cache/) but can't complete it without your help, as it would need a new S3 bucket to store cache in:
> 
> ```
> level=error name=drone-cache ts=2021-05-23T20:43:54.270125703Z caller=main.go:466 err="[WARNING] build cache, process rebuild failed, upload upload file put the object NoSuchBucket: The specified bucket does not exist\n\tstatus code: 404, request id: HK5NZ5DCKE9VHRJT, host id: QfOewvsVfzFb8hsil1/tdyQk4T1bJPXLgT0czTa39Uwl3/9NwsxOXJRw91TFAAHOjxM00YJZmL0=\n"
> ```

It takes a way more cpu cycles to compile MCS for sure than the server itself. Let's see if drone cache could reduce the timings.

**Pull request closed without merging**